libs/agno/agno/knowledge/embedder/cohere.py

```diff
@@ -61,6 +64,189 @@ def response(self, text: str) -> Union[EmbeddingsFloatsEmbedResponse, Embeddings
+        elif attempt == 1:
+            delay = 30.0  # Wait 30 seconds (1/2 of minute window)
+        else:
+            delay = 60.0  # Wait full minute for window reset
+
+        # Add small jitter
+        delay += time.time() % 3
+
+        log_debug(
+            f"Async rate limit backoff, waiting {delay:.1f} seconds for rate limit window reset (attempt {attempt + 1})"
+        )
+        await asyncio.sleep(delay)
+
+    def _batch_with_retry(
+        self, texts: List[str], max_retries: int = 3
+    ) -> Tuple[List[List[float]], List[Optional[Dict]]]:
+        """Execute batch embedding with rate-limit-aware backoff for rate limiting."""
+        log_debug(f"Starting batch retry for {len(texts)} texts with max_retries={max_retries}")
+
+        for attempt in range(max_retries + 1):
+            try:
+                request_params = self._get_batch_request_params()
+                response: Union[EmbeddingsFloatsEmbedResponse, EmbeddingsByTypeEmbedResponse] = self.client.embed(
+                    texts=texts, **request_params
+                )
+
+                # Extract embeddings from response
+                if isinstance(response, EmbeddingsFloatsEmbedResponse):
+                    batch_embeddings = response.embeddings
+                elif isinstance(response, EmbeddingsByTypeEmbedResponse):
+                    batch_embeddings = response.embeddings.float_ if response.embeddings.float_ else []
+                else:
+                    log_warning("No embeddings found in response")
+                    batch_embeddings = []
+
+                # Extract usage information
+                usage = response.meta.billed_units if response.meta else None
+                usage_dict = usage.model_dump() if usage else None
+                all_usage = [usage_dict] * len(batch_embeddings)
+
+                log_debug(f"Batch embedding succeeded on attempt {attempt + 1}")
+                return batch_embeddings, all_usage
+
+            except Exception as e:
+                if self._is_rate_limit_error(e):
+                    if not self.exponential_backoff:
+                        log_warning(
+                            "Rate limit detected. To enable automatic backoff retry, set enable_backoff=True when creating the embedder."
+                        )
+                        raise e
+
+                    log_info(f"Rate limit detected on attempt {attempt + 1}")
+                    if attempt < max_retries:
+                        self._rate_limit_backoff_sleep(attempt)
+                        continue
+                    else:
+                        log_warning(f"Max retries ({max_retries}) reached for rate limiting")
+                        raise e
+                else:
+                    log_debug(f"Non-rate-limit error on attempt {attempt + 1}: {e}")
+                    raise e
+
+        # This should never be reached, but just in case
+        log_error("Unexpected: Reached end of retry loop")
```

Comment:
The error is kinda vague.
```suggestion
        log_error("Could not create embeddings. End of retry loop reached.")
```